pkg/library/gatewayapi/topology.go

```diff
@@ -18,81 +18,192 @@ const (
+	// get children of Policy kind
+	policyNodeList := utils.Filter(g.graph.Children(g.ID()), func(n dag.Node) bool {
+		_, ok := n.(policyDAGNode)
+		return ok
+	})
 
-	routes []RouteNode
+	return utils.Map(policyNodeList, func(n dag.Node) PolicyNode {
+		return PolicyNode{n.(policyDAGNode), g.graph}
+	})
 }
 
-func (g *GatewayNode) AttachedPolicies() []Policy {
-	return g.attachedPolicies
+func (g *GatewayNode) GetGateway() *gatewayapiv1.Gateway {
+	return g.Gateway
 }
 
 func (g *GatewayNode) Routes() []RouteNode {
-	return g.routes
+	// get children of httproute kind
+	routeNodeList := utils.Filter(g.graph.Children(g.ID()), func(n dag.Node) bool {
+		_, ok := n.(httpRouteDAGNode)
+		return ok
+	})
+
+	return utils.Map(routeNodeList, func(n dag.Node) RouteNode {
+		routeDAGNode := n.(httpRouteDAGNode)
+		return RouteNode{routeDAGNode, g.graph}
+	})
 }
 
 func (g *GatewayNode) ObjectKey() client.ObjectKey {
 	return client.ObjectKeyFromObject(g.Gateway)
 }
 
+func (g *GatewayNode) GetGatewayNode() *GatewayNode {
+	return g
+}
+
+func (g *GatewayNode) GetRouteNode() *RouteNode {
+	return nil
+}
+
+// Topology defines a graph with Gateway API entities.
+// Contains GatewayNodes (Gateway API gateways)
+// Contains RouteNodes (Gateway API httproutes)
+// Contains PolicyNodes (Gateway API policy attachment objects)
+// Hierarchy is as follows.
+// GatewayNode children can be either RouteNode or PolicyNode nodes
+// RouteNode children are PolicyNode nodes
+// PolicyNode parents can be either RouteNode or GatewayNode nodes
 type Topology struct {
 	graph  *dag.DAG
 	Logger logr.Logger
 }
 
 type gatewayDAGNode struct {
 	*gatewayapiv1.Gateway
-
-	attachedPolicies []Policy
 }
 
 func dagNodeIDFromObject(obj client.Object) dag.NodeID {
 	return fmt.Sprintf("%s#%s", obj.GetObjectKind().GroupVersionKind().String(), client.ObjectKeyFromObject(obj).String())
 }
 
-func (g gatewayDAGNode) ID() string {
+func (g gatewayDAGNode) ID() dag.NodeID {
 	return dagNodeIDFromObject(g.Gateway)
 }
 
 type httpRouteDAGNode struct {
 	*gatewayapiv1.HTTPRoute
-
-	attachedPolicies []Policy
 }
 
-func (h httpRouteDAGNode) ID() string {
+func (h httpRouteDAGNode) ID() dag.NodeID {
 	return dagNodeIDFromObject(h.HTTPRoute)
 }
 
+type policyDAGNode struct {
+	Policy
+}
+
+func (p policyDAGNode) ID() dag.NodeID {
+	return dagNodeIDFromObject(p.Policy)
+}
+
 type topologyOptions struct {
-	gateways []*gatewayapiv1.Gateway
-	routes   []*gatewayapiv1.HTTPRoute
-	policies []Policy
-	logger   logr.Logger
+	gateways               []*gatewayapiv1.Gateway
+	routes                 []*gatewayapiv1.HTTPRoute
+	policies               []Policy
+	logger                 logr.Logger
+	programmedGatewaysOnly bool
+	linkAcceptedRoutesOnly bool
 }
 
 // TopologyOpts allows to manipulate topologyOptions.
 type TopologyOpts func(*topologyOptions)
 
+func WithAcceptedRoutesLinkedOnly() TopologyOpts {
+	return func(o *topologyOptions) {
+		o.linkAcceptedRoutesOnly = true
```

Comment:
I guess we don't ever want non-accepted routes in the topology as we may occasionally want non-programmed gateways? I actually wonder what would be the use case for adding non-programmed gateways.

Reply:
there are some use cases I found that non-accepted routes should be linked to parent gateways: mappers. It allows getting events for "disconnected" policies, which status can be affected from the "disconnection".

Reply:
>  I actually wonder what would be the use case for adding non-programmed gateways.

this is a use case https://github.com/Kuadrant/kuadrant-operator/pull/751#discussion_r1671913835

Reply:
> there are some use cases I found that non-accepted routes should be linked to parent gateways

So why not giving the option here then?

Reply:
Sorry. To clarify...

My point is that we have `WithProgrammedGatewaysOnly(bool)` for when we want to enforce programmed gateways only, i.e. `WithProgrammedGatewaysOnly(true)`. I don't see why one would use `WithProgrammedGatewaysOnly(false)` though – versus simply not calling `WithProgrammedGatewaysOnly` at all.

I guess that was your thinking for `WithAcceptedRoutesLinkedOnly`. Nobody will ever need `WithAcceptedRoutesLinkedOnly(false)`. The effect in this case is virtually the same as not using `WithAcceptedRoutesLinkedOnly`, right?

So why not designing `WithProgrammedGatewaysOnly` the same way? If there's a reason I struggling to see, then there may be for `WithAcceptedRoutesLinkedOnly` too perhaps?

IOW, why are the signatures of these 2 functions different?

Reply:
Makes all the sense. Fixing it

Reply:
done